File: src/test/java/com/woomulwoomul/woomulwoomulbackend/domain/user/UserRoleRepositoryTest.java
package com.woomulwoomul.woomulwoomulbackend.domain.user;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static com.woomulwoomul.woomulwoomulbackend.domain.user.ProviderType.KAKAO;
import static com.woomulwoomul.woomulwoomulbackend.domain.user.Role.USER;
import static org.assertj.core.api.Assertions.assertThat;

@ActiveProfiles("test")
@SpringBootTest
@Transactional
class UserRoleRepositoryTest {

    @Autowired private UserRoleRepository userRoleRepository;
    @Autowired private UserRepository userRepository;
    @Autowired private UserProviderRepository userProviderRepository;

    @DisplayName("회원 권한 조인 회원 전체 조회가 정상 작동한다")
    @Test
    void givenValid_whenFindAllFetchUser_thenReturn() {
        // given
        UserRole userRole = createAndSaveUserRole();

        // when
        List<UserRole> userRoles = userRoleRepository.findAllFetchUser(userRole.getUser().getId());

        // then
        assertThat(userRoles).containsExactly(userRole);
    }

    private UserRole createAndSaveUserRole() {
        UserProvider userProvider = userProviderRepository.save(UserProvider.builder()
                .provider(KAKAO)
                .providerId("1")
                .build());

        User user = userRepository.save(User.builder()
                .userProvider(userProvider)
                .username("tester")
                .imageUrl("")
                .build());

        return userRoleRepository.save(UserRole.builder()
                .user(user)
                .role(USER)
                .build());
    }
}

File: src/test/kotlin/com/woomulwoomul/woomulwoomulbackend/domain/user/UserRoleRepositoryTest.kt
package com.woomulwoomul.woomulwoomulbackend.domain.user

import com.woomulwoomul.woomulwoomulbackend.domain.user.Role.USER
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.groups.Tuple
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

@ActiveProfiles("test")
@SpringBootTest
@Transactional
class UserRoleRepositoryTest(
    @Autowired private val userRoleRepository: UserRoleRepository,
    @Autowired private val userRepository: UserRepository
) {

    @DisplayName("회원 ID로 모든 회원 권한을 조회하면 정상 작동한다")
    @Test
    fun givenValid_whenFindAllFetchUser_thenReturn() {
        // given
        val userRole = createAndSaveUserRole(USER)

        // when
        val foundUserRoles = userRoleRepository.findAllFetchUser(userRole.user.id!!)

        // then
        assertAll(
            {
                assertThat(foundUserRoles)
                    .extracting("id", "role", "status", "createDateTime", "updateDateTime")
                    .containsExactly(
                        Tuple.tuple(userRole.id, userRole.role, userRole.status,
                            userRole.createDateTime, userRole.updateDateTime)
                    )
            },
            {
                assertThat(foundUserRoles)
                    .extracting("user")
                    .extracting("id", "username", "imageUrl", "status", "createDateTime", "updateDateTime")
                    .containsExactly(
                        Tuple.tuple(userRole.user.id, userRole.user.username,
                            userRole.user.imageUrl, userRole.user.status,
                            userRole.user.createDateTime, userRole.user.updateDateTime)
                    )
            }
        )
    }

    private fun createAndSaveUserRole(role: Role): UserRoleEntity {
        val user = userRepository.save(UserEntity(
            username = "tester",
            email = "tester@woomulwoomul.com",
            imageUrl = "https://t1.kakaocdn.net/account_images/default_profile.jpeg.twg.thumb.R640x640"
        ))
        return userRoleRepository.save(UserRoleEntity(user = user, role = role))
    }
}